File: src/providers/openai-realtime/rules/reconnect-on-drop.ts
import { collectOpenAIRealtimeSocketVarNames, isSocketOnCall } from '../utils.js';

const RECONNECT_NAME_PATTERN = /reconnect/i;

/** True when `subtree` contains a `new WebSocket(...)` or a call to something named like "reconnect". */
function hasReconnectAttempt(node: any, depth = 0): boolean {
  if (!node || typeof node !== 'object' || depth > 60) return false;
  if (Array.isArray(node)) {
    return node.some((n) => hasReconnectAttempt(n, depth + 1));
  }

  if (node.type === 'NewExpression' && node.callee?.type === 'Identifier' && node.callee.name === 'WebSocket') {
    return true;
  }

  if (node.type === 'CallExpression') {
    const callee = node.callee;
    const calleeName =
      callee?.type === 'Identifier'
        ? callee.name
        : callee?.type === 'MemberExpression' && callee.property?.type === 'Identifier'
          ? callee.property.name
          : null;
    if (typeof calleeName === 'string' && RECONNECT_NAME_PATTERN.test(calleeName)) return true;
  }

  for (const key of Object.keys(node)) {
    if (key === 'parent' || key === 'loc' || key === 'range') continue;
    const val = node[key];
    if (val && typeof val === 'object' && hasReconnectAttempt(val, depth + 1)) return true;
  }
  return false;
}

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'A dropped OpenAI Realtime connection must be retried, not just logged',
      category: 'reliability',
      docsUrl: 'https://developers.openai.com/api/docs/guides/realtime',
      rationale:
        "For a multi-minute phone call, a single transient OpenAI-side disconnect that is only logged on 'close' permanently kills translation/processing for the remainder of the call in that direction — the call itself stays connected and silently degraded, with neither party getting a signal that something stopped working.",
      recommended: true,
    },
    messages: {
      noReconnectAttempt:
        "This Realtime socket's close handler only logs and never attempts to reconnect.",
    },
    schema: [],
  },
  create(context: any) {
    let socketVarNames = new Set<string>();

    return {
      Program(node: any) {
        socketVarNames = collectOpenAIRealtimeSocketVarNames(node);
      },

      CallExpression(node: any) {
        if (socketVarNames.size === 0) return;
        if (!isSocketOnCall(node, socketVarNames, 'close')) return;

        const handler = node.arguments?.[1];
        if (handler?.type !== 'ArrowFunctionExpression' && handler?.type !== 'FunctionExpression') return;

        if (!hasReconnectAttempt(handler.body)) {
          context.report({ node, messageId: 'noReconnectAttempt' });
        }
      },
    };
  },
};

export const openaiRealtimeReconnectOnDropRule = rule;

File: src/providers/openai-realtime/rules/python/reconnect_on_drop.py
"""openai-realtime-reconnect-on-drop

Parity with the JS rule: an except handler for a dropped Realtime connection
(`websockets.exceptions.ConnectionClosed` and friends) that only logs and
never attempts to reconnect permanently kills the call in that direction.
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (  # noqa: E402
    collect_realtime_socket_var_names,
    collect_realtime_url_var_names,
    is_realtime_connect_call,
    is_tracked_socket_ref,
)

RULE_KEY = "openai-realtime-reconnect-on-drop"

RECONNECT_NAME_RE = re.compile(r"reconnect", re.I)
CONNECTION_CLOSED_RE = re.compile(r"connectionclosed", re.I)


def _handler_type_names(expr: ast.AST | None) -> list[str]:
    if expr is None:
        return []
    if isinstance(expr, ast.Tuple):
        names: list[str] = []
        for el in expr.elts:
            names.extend(_handler_type_names(el))
        return names
    if isinstance(expr, ast.Attribute):
        return [expr.attr]
    if isinstance(expr, ast.Name):
        return [expr.id]
    return []


def _handler_matches_connection_closed(handler: ast.ExceptHandler) -> bool:
    return any(CONNECTION_CLOSED_RE.search(n) for n in _handler_type_names(handler.type))


def _has_reconnect_attempt(nodes: list[ast.stmt], url_var_names: set[str]) -> bool:
    for stmt in nodes:
        for n in ast.walk(stmt):
            if isinstance(n, ast.Call):
                if is_realtime_connect_call(n, url_var_names):
                    return True
                name = n.func.id if isinstance(n.func, ast.Name) else (
                    n.func.attr if isinstance(n.func, ast.Attribute) else None
                )
                if name and RECONNECT_NAME_RE.search(name):
                    return True
    return False


def _try_touches_realtime(node: ast.Try, url_var_names: set[str], socket_var_names: set[str]) -> bool:
    for stmt in node.body:
        for n in ast.walk(stmt):
            if isinstance(n, ast.Call) and is_realtime_connect_call(n, url_var_names):
                return True
            if is_tracked_socket_ref(n, socket_var_names):
                return True
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    url_var_names = collect_realtime_url_var_names(tree)
    socket_var_names = collect_realtime_socket_var_names(tree)
    if not url_var_names and not socket_var_names:
        return []

    out: list[dict] = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Try):
            continue
        if not _try_touches_realtime(node, url_var_names, socket_var_names):
            continue

        for handler in node.handlers:
            if not _handler_matches_connection_closed(handler):
                continue
            if _has_reconnect_attempt(handler.body, url_var_names):
                continue
            line, col, _, _ = loc(handler)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
